### PR TITLE
Add protocol to define methods relied upon by KubernetesPodOperator

### PR DESCRIPTION
**Pull request by dstandish**
Subclasses of KubernetesPodOperator, such as GKEStartPodOperator, may use hooks that don't extend KubernetesHook.  We use this protocol to document the methods used by KPO and ensure that these methods exist on such other hooks.

Related: https://github.com/apache/airflow/pull/31266

In https://github.com/apache/airflow/pull/31266, I added the missing methods to GKESPO and test to ensure consistency.  With the introduction of this protocol, this test would no longer be needed, though it will require a dependency for google on the latest CNCF provider

### CODE REVIEW

**Review by potiuk — APPROVED** [review]

**Review by hussein-awala — APPROVED** [review]
Nice one

**Comment by potiuk** [comment]
> @eladkal @potiuk I am surprised to not see cncf in the list of dependencies for google provider, because the GKPO does inherit from KPO.
> 

It does, implicitly. https://github.com/apache/airflow/blob/d3a2dbe028e4a8c2da9fcb2b3257e2117c413a73/generated/provider_dependencies.json#L394 

When we prepare "generated/provider_dependencies.json", the cross-provider-dependencies are automatically derived by looking at teh imports of other providers from a provider - and they are automatically added to the dependencies and "extras" of the provider:

Cross provider package dependencies as "extras" https://airflow.apache.org/docs/apache-airflow-providers-google/stable/index.html#cross-provider-package-dependencies 

For example in this case `pip install apache-airflow-providers-google[cncf.kubernetes]` will automatically install the cncf.kubernetes provider.

This is a "weak" relation between providers which means that some things in the provider **might** not work in case the extra is not used when installing the provider (or when by chance the provider from extra is not installed). But if we think that the dependency is so strong that one cannot live by another - we can add explicit dependency (and even add >= explicit version which happened in some cases) in the past.

> Given that with this we'd be introducing a dep in GKPO that depends on a protocol defined in CNCF, first of all is this even workable? And how should we handle this with regard to deps?

Given that this is one of about a 160 or so operators in google provider, I'd say the weak relation we have is "good enough". You can make an extremely good use of the  159 or so operators in the google provider, without having cncf.kubernetes one installed.

**Comment by potiuk** [comment]
Btw - maybe thi is what you are after - we could -in fact - make even explicit VERSIONED extra dependency.

This is already happening for google provider `provider.yaml':

```
additional-extras:
  - name: apache.beam
    dependencies:
      - apache-beam[gcp]
  - name: leveldb
    dependencies:
      - plyvel
  - name: oracle
    dependencies:
      - apache-airflow-providers-oracle>=3.1.0
  - name: facebook
    dependencies:
      - apache-airflow-providers-facebook>=2.2.0
  - name: amazon
    dependencies:
      - apache-airflow-providers-amazon>=2.6.0
```

You could add in :

```
  - name: cncf.kubernetes
    dependencies:
      - apache-airflow-providers-cncf-kubernetes>=6.2.0
```

And what would happen, if someone does `pip install apache-airflow-providers-google[cncf.kubernetes]` - then minimum 6.2.0 version of the cncf.kubernetes provider will be installed (assuming that the protocol is going to be added in 6.2.0).

File: airflow/providers/cncf/kubernetes/utils/pod_manager.py [hunk 1]
```diff
@@ -24,7 +24,7 @@
 from contextlib import closing, suppress
 from dataclasses import dataclass
 from datetime import datetime, timedelta
-from typing import TYPE_CHECKING, Generator, cast
+from typing import TYPE_CHECKING, Generator, Protocol, cast
```
Comment (potiuk) [inline airflow/providers/cncf/kubernetes/utils/pod_manager.py]: Small problem - until end of June, Protocol needs to be imported as:

```
from airflow.typing_compat import Protocol
```

Comment (dstandish) [reply airflow/providers/cncf/kubernetes/utils/pod_manager.py]: thanks

**Comment by dstandish** [comment]
> You could add there:
> 
> ```
>   - name: cncf.kubernetes
>     dependencies:
>       - apache-airflow-providers-cncf-kubernetes>=6.2.0
> ```
> 
> And what would happen, if someone does `pip install apache-airflow-providers-google[cncf.kubernetes]` - then minimum 6.2.0 version of the cncf.kubernetes provider will be installed (assuming that the protocol is going to be added in 6.2.0).

ok thanks i will do this

**Comment by potiuk** [comment]
> > And what would happen, if someone does `pip install apache-airflow-providers-google[cncf.kubernetes]` - then minimum 6.2.0 version of the cncf.kubernetes provider will be installed (assuming that the protocol is going to be added in 6.2.0).

Yep. Just checked in https://github.com/apache/airflow/pull/31252/files that we are going to have 6.2.0 version of the cncf.kubernetes provider.

**Comment by uranusjr** [comment]
I’m trying to bulk-fix the migration errors in #31306.

**Comment by dstandish** [comment]
> I’m trying to bulk-fix the migration errors in #31306.

cool, will wait for that one

i tried to fix them all in https://github.com/apache/airflow/pull/31306 but it got merged just before i discovered there were others and tried to add them

**Comment by dstandish** [comment]
er i'll just merge, don't think it will create a conflict for you

**Pull request merged**